subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java

```diff
@@ -20,11 +20,14 @@
 
 import java.io.File;
 import java.util.List;
+import java.util.Map;
 
 public interface CompilerMetaDataProvider<T extends CompilerMetadata> {
 
     SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath);
 
+    SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath, Map<String, String> environmentVariables);
```

Comment:
❓ This is by far the ugliest part of this PR. Any idea how to improve this? To switch between Xcode installation, the process needs to have `DEVELOPER_DIR` set to the installation, `.app`, directory. It's the official method, `xcrun` trick requires root access. We don't have any modelling within Gradle for Xcode to specify the `installDir`.

It's internal API, however I almost feel like we should be using a `Action<ExecSpec>` instead but I don't know how we could cache the call (inside `CachingCompilerMetaDataProvider`). Also, we don't allow all `ExecSpec` method to be configured. We could introduce a `CompilerExecSpec` that exposes a subset of the `ExecSpec` methods that allow us to extract the values and cache them. Something like:
```
interface CompilerExecSpec {
  CompilerExecSpec environment(String key, String value);
  CompilerExecSpec executable(Object executable);
  CompilerExecSpec args(String... args);
}
```
The prototype would become:
```
SearchResult<T> getCompilerMetaData(List<File> searchPath, Action<? super CompilerExecSpec> configureAction);
```